BUG: signal: Control output dtypes in `abcd_normalize` by adding new …

DietBru opened the pull request:
With this PR `signal.abcd_normalize` gains the capability of controlling the dtype of the returned arrays through the new keyword parameter ~~`cast_to`~~ `dtype`. The new default behavior is to make the output dtype float64 / complex128 (depending on input dtypes). Alternatively, either a specific dtype can be provided or type conversion can be turned off.

The idea behind adding a new keyword parameter was to make it clearly visible what `abcd_normalize` is actually doing.

Additionally:
* `signal.ss2tf`, `signal.StateSpace` now convert parameters to float64 / complex128 due to relying on `abcd_normalize`.
* ~~The initializer of class `signal.StateSpace` gained the keyword `cast_to`, which is passed to `abcd_normalize` for normalizing matrix parameters. Hence, those are converted to floating by default.~~
* Unit tests were added to ensure 100% coverage.
* Marginal docstr improvements.
* Supersedes PR #13235 from 2020: The `abcd_normalize` code diverged too much for making sensible adaptions in that PR.
* Closes #13125.
* Closes #18982.
* Closes #19629.

DietBru committed:
With this commit `signal.abcd_normalize` gains a new keyword `dtype` to control the dtypes of the returned matrices. The new default behavior (`dtype=None) is to make the output dtype float64 / complex128 (depending on input dtypes) instead of preserving the dtype.

* `signal.ss2tf` and the initializer of class `Statespace` now convert parameters to floating due to relying on `abcd_normalize`.
* Add Array API compatbility.
* Unit tests were added to ensure 100% coverage.
* Marginal docstr improvements.
* Supersedes PR 13235 from 2020: The `abcd_normalize` code diverged too much for making sensible adaptions in that PR.
* Closes issue 13125, closes issue 18982.

DietBru commented:
Rendered docs: [abcd_normalize], [ss2tf], [StateSpace]


[abcd_normalize]: https://output.circle-artifacts.com/output/job/a57774fa-3db0-4946-a13a-1f30021b0872/artifacts/0/html/reference/generated/scipy.signal.abcd_normalize.html
[ss2tf]: https://output.circle-artifacts.com/output/job/a57774fa-3db0-4946-a13a-1f30021b0872/artifacts/0/html/reference/generated/scipy.signal.ss2tf.html#scipy.signal.ss2tf
[StateSpace]: https://output.circle-artifacts.com/output/job/a57774fa-3db0-4946-a13a-1f30021b0872/artifacts/0/html/reference/generated/scipy.signal.StateSpace.html

ilayn commented on scipy/signal/_ltisys.py:
```diff
@@ -1242,6 +1242,13 @@ class StateSpace(LinearTimeInvariant):
         Sampling time [s] of the discrete-time systems. Defaults to `None`
         (continuous-time). Must be specified as a keyword argument, for
         example, ``dt=0.1``.
+    cast_to: dtype | 'floating64' | None, optional
```
Everytime we do this, after 5-6 years it becomes an issue when something needs to change or a deprecation comes. If there is no use case for this, I'd suggest we don't bubble up this functionality to `StateSpace`. It's much easier to request from the user to use proper dtype for it. Integers to floats are expected to be done automatically elsewhere. StateSpace function API is not the right place to do data casting unless we implement the entire NumPy casting machinery.

DietBru replied:
Removed in commit 3708d92. If needed, it can be easily added again.

DietBru commented on scipy/signal/_lti_conversion.py:
```diff
@@ -134,6 +134,10 @@ def abcd_normalize(A=None, B=None, C=None, D=None):
         Two-dimensional array of shape (q, n).
     D: array_like, optional
         Two-dimensional array of shape (q, p).
+    cast_to: dtype | 'floating64' | None, optional
```
You're right, having unified dtypes are probably easier to handle—done in commit 3708d92.

I still can think of some esoteric use cases, where keeping the input dtype is beneficial—if needed a case of `dtype='keep'` can still be added again.

I also added the keyword parameters `xp` and `device` in commit 3708d92 to work around some dtype oddities in the unit tests (if A, B, C, D are all lists, "guessing" the "xp"-namespace is not always correct).

DietBru commented:
Rendered docs (update): [abcd_normalize], [ss2tf], [StateSpace].




[abcd_normalize]: https://output.circle-artifacts.com/output/job/f4d6e682-4d0e-448e-b399-2c8afb2e1bae/artifacts/0/html/reference/generated/scipy.signal.abcd_normalize.html
[ss2tf]: https://output.circle-artifacts.com/output/job/f4d6e682-4d0e-448e-b399-2c8afb2e1bae/artifacts/0/html/reference/generated/scipy.signal.ss2tf.html#scipy.signal.ss2tf
[StateSpace]: https://output.circle-artifacts.com/output/job/f4d6e682-4d0e-448e-b399-2c8afb2e1bae/artifacts/0/html/reference/generated/scipy.signal.StateSpace.html

ev-br reviewed (changes requested):
I've to admit I'm a bit confused. There seems to be two things going on here: fix of how integer dtypes are handled, and the array api conversion? It would be easier to follow if these two things are separate into array api conversion (+ strong backwards compat for array-like inputs) and a dtype fix---either as two PRs or as clearly separate commits. Would it be possible to split this PR (for me, either separate PRs or separate commits are fine, so whichever is more convenient to you).
I left a couple of comments mostly relating to the former part, the array api conversion.

For the dtype fix, I am also not clear why there is a need for the new `dtype=` kwarg actually. If the problem is that integer arrays do not get properly converted to floats, would it work to just convert them internally?   Naively---and it's full well possible I'm missing something---would https://github.com/scipy/scipy/blob/main/scipy/interpolate/_bsplines.py#L23-L41 work?

lucascolley commented:
are you interested in getting back to this for 1.17.0 @DietBru ?

DietBru committed:
Fix `spin smoke-docs` in signal module.

Makes `spin smoke-docs -s signal` process without errors.

The docstrs of the following functions where changed:

* scipy.signal_filter_design.noramlize - prevent exception by warning.

* scipy.signal._filter_design._bessel_poly - change to code block due to large output.

* In `scipy.signal._ltisys`: `TransferFunctionContinuous`, `TransferFunctionDiscrete`, `StateSpaceContinuous`, `StateSpaceDiscrete`

* scipy.signal._signaltools._cmplx_sort

DietBru commented on scipy/signal/_lti_conversion.py:
```diff
@@ -181,27 +191,67 @@ def abcd_normalize(A=None, B=None, C=None, D=None):
-    A, B, C, D = (xpx.atleast_nd(xp.asarray(M_), ndim=2) if M_ is not None else
-                  xp.zeros((0, 0)) for M_ in (A, B, C, D))
+    if xp is None:  # try to determine namespace if not given as parameter:
+        xp = array_namespace(A, B, C, D)
```
Fixed in bd776702da5bbabdb99c6056e939dd9e89f4415d. I guess I got confused between the differences of `numpy.asarray` and `np_compat.asarray`.

ev-br commented on scipy/signal/tests/test_cont2discrete.py:
```diff
@@ -359,8 +359,8 @@ class TestC2dLti:
-        B = np.array([[0], [1]])
-        C = np.array([[1, 0]])
+        B = np.array([[0.], [1.]])
+        C = np.array([[1., 0.]])
```
Do I understand it right that the code change is backwards compatible: if `A` was integer, after `abcd_normalize` it will become float64? If so, the test does not need changing.

ev-br reviewed (commented):
A couple of minor comments below. The main question IMO is about mixing array arguments and an explicit `xp=` argument: what is the target behavior for `A=numpy_array` and `xp=torch`? I don't know the answer, and I actually don't know if there's a compelling reason to allow this (and once added, it's hard to remove something). My gut feeling is to not add it, but it's not my hill and I won't press the point if domain experts consider it desirable in this particular function.

lucascolley commented on scipy/signal/_lti_conversion.py:
```diff
@@ -114,44 +114,66 @@ def tf2ss(num, den):
+    xp : array_namespace, optional
+        The namespace for the return array. When ``None`` (default), then it is derived
+        from the parameters `A`, `B`, `C`, `D`. It should be compatible with the array
+        API standard or supported by array-api-compat. ``numpy`` is used by if
```
Agreed, the pattern followed elsewhere has been to infer from the input whenever there is an array parameter. I'm not aware of compelling reasons to support e.g. passing in scalars and getting back a torch array.

DietBru replied:
So be it. Keyword `device` is removed in commit a7fc9b038a368a1bccd740677e81523638b4ab66.

lucascolley replied:
> My impression is that we either want both `xp=..., device=...` or neither of them.

Yes, that is my impression too

DietBru committed:
For PR #23554:
* In `test_cont2discrete.TestC2dLtitest:2d:ss`: Changed input dtypes of B,D back to int
* Added back `abcd_normalize_signature` to `signal._delegators.py`.
* Minor improvement in example of docstr of `signal.abcd_normalize`

DietBru commented on scipy/signal/_lti_conversion.py:
```diff
@@ -181,6 +203,21 @@ def abcd_normalize(A=None, B=None, C=None, D=None):
+    >>> print(f" AA: {AA.dtype}, BB: {BB.dtype}\n CC: {CC.dtype}, DD: {DD.dtype}")
+     AA: float64, BB: float64
+     CC: float64, DD: float64
+    >>> AA, BB, CC, DD = abcd_normalize(A=A, D=D, dtype=np.int32)  # Explicit dtype
```
Agreed—fixed in cf1ce6a93501e337e273e90c3719e01252f43c72

DietBru committed:
Add `signal.abcd_normalize` to `signal._support_alternative_backends.CUPY_BLACKLIST` due to failing CI test.

Furthermore, `signal._support_alternative_backends.CUPY_BLACKLIST` was sorted into alphabetical order.

This commit is part of PR #23554.

DietBru committed:
Remove keyword `xp` from `signal.abcd_normalize`

This commit is part of PR #23554.

ev-br commented on scipy/signal/_delegators.py:
```diff
@@ -58,10 +58,6 @@ def _skip_if_poly1d(arg):
 ###################
 
-def abcd_normalize_signature(A=None, B=None, C=None, D=None):
-    return array_namespace(A, B, C, D)
```
Ah, it's probably the `xp-capabilities` dance in [`_support_alternative_backends.py`](https://github.com/scipy/scipy/blob/main/scipy/signal/_support_alternative_backends.py) makes an assumption that all delegators exist, and does not apply them if it is not.
This is literally under two weeks old, so if it gets in a way, we totally can fix it. Up to you, I suppose.

DietBru committed:
Remove keyword `device` from `signal.abcd_normalize`

This commit is part of PR #23554.

DietBru committed:
Remove decorator from `xp_capabilities` from `signal.abcd_normalize`.

This commit is part of PR #23554.

ev-br reviewed (approved):
LGTM now. The circle CI failures are unrelated, and are fixed now in main, so probably safe to ignore.

Thank you @DietBru , it's great to see this part of scipy getting more dev attention!

DietBru commented:
> I agree, I just think that adding a dtype kwarg and not respecting input dtype doesn't offer any possibilities that can't be done by either respecting input dtype, or using xp.astype.

Well, `abcd_normalize` [uses]  `xp.astype`  to ensure identical output dtypes—so why not expose that parameter?  Maybe renaming the parameter `dtype` to `astype` would make that clearer. The difference in our views probably boils down to the perception on how the state space functionality gets typically used. My guess would be that short semi-interactive scripts with NumPy for educational purposes are the main use case. However, it's probably not unfair to assume that the state space stuff does not see too much use. For that reason I think that having or not having  the `dtype` parameter would both be okay.

> Also, are you aware of [mparray](https://github.com/mdhaber/mparray)?

Very interesting! I was not aware of it—I'll definitely have a closer look. Thanks.

> testing with different dtypes should be explicitly done by developers.

:+1: I fully agree with that. 

Thank you for taking the time for explaining your thoughts in detail. 
#24130 is a great improvement over the status quo and in the right direction.


[uses]: https://github.com/scipy/scipy/blob/95af6337fff1f49bb10c99bca3b4f40802335305/scipy/signal/_lti_conversion.py#L246

steppi commented:
Cheers @ilayn! Thanks for weighing in. I don't think we disagree as much as you think, so I'll try to lay everything out clearly. I was guilty of writing things too quickly. I think if I'd actually been trying to say what @DietBru thought I was trying to say then I'd deserve a few nice words. 

I don't actually think it should be a requirement that SciPy should always be consistent about preserving input `dtype`. I know I wrote that above, and probably sounded blindly dogmatic, but what I was really trying to say is that if the goal is to provide a way to control output `dtype` and allow for native computations in the `dtype` of choice, then respecting input `dtype`s with typical promotion rules in my opinion offers a clean API which is straightforward to implement. https://github.com/scipy/scipy/pull/24232 should make `abcd_normalize` easier to maintain  my opinion. If one doesn't care about the function allowing native computation in the `dtype` of choice, wants it to always upcast because single precision can't be trusted, but still cares about changing output `dtype`, then anything one can do with the `dtype` kwarg could also be accomplished by applying `astype` after the fact. If you thinks its better if `abcd_normalize` always upcasts to `float64`, I'd be fine with that too so long as its documented.

I was told the same thing about a `dtype` kwarg not being needed if you can just use `astype` after the fact in the review for this PR, https://github.com/scipy/scipy/pull/23921, where I was taking the opposite side and wanted to add a `dtype` kwarg to filter prototype functions. @ev-br and I ended up just making all filter prototype functions return double precision outputs regardless of input type, for pretty much the same reasons your saying. I documented that this might change in the future, but I don't care if these functions keep the current behavior.

PyTorch arrays probably never should have become a core part of this discussion. My opinion on those is that supporting PyTorch shouldn't have to change how SciPy developers think about dtypes. Just because there are tests with PyTorch having default dtype float32, doesn't mean SciPy devs need to start adding `float32` support for everything to get those tests to pass. If one blindly converts tests to support the array API, then that's what the inevitable failures you get with the torch float32 tests seem to be saying. @DietBru had mentioned adding the `dtype` kwarg because of the torch float32 failures, and I was just trying to say, "no, you shouldn't worry about those at all", and was trying to speak from my experience trying to add an unnecessary `dtype` kwarg to filter prototype functions precisely to make it easier to get those `float32` tests to pass. But things came out wrong and I emphasized the wrong things.

ilayn commented:
Yes I think there is no disagreement indeed. What I am trying to emphasize is that we made a decision and many changes we are doing are due to that commitment. We can hold on to that argumentation for the reasoning, and we don't need to invoke some type consistency argument or some type-theoretical argument to justify it. Because then it opens a type discussion which is not going to help because we already made the decision. 

We can save quite a bit of friction and kilobytes of text I believe if we can avoid that. Other than that, all good.
